shanoir-ng-studies/src/main/java/org/shanoir/ng/study/service/RelatedDatasetServiceImpl.java

```diff
@@ -0,0 +1,173 @@
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Implementation of study service.
+ * 
+ * @author msimon
+ *
+ */
+@Component
+public class RelatedDatasetServiceImpl implements RelatedDatasetService {
+
+	@Autowired
+	private StudyUserRepository studyUserRepository;
+	@Autowired
+	private StudyService studyService;
+	@Autowired
+	private CenterRepository centerRepository;
+	@Autowired
+	private StudyRepository studyRepository;
+	@Autowired
+	private SubjectRepository subjectRepository;
+	@Autowired
+	private RabbitTemplate rabbitTemplate;
+	private static final Logger LOG = LoggerFactory.getLogger(RelatedDatasetServiceImpl.class);
+
+	@Transactional
+	@Override
+	public void addSubjectStudyToNewStudy(String subjectIds, String id) {
+		Study study = studyService.findById(Long.valueOf(id));
+		Boolean toAdd = true;
+		for (String subjectId : subjectIds.split(",")) {
+			Subject subject = subjectRepository.findById(Long.valueOf(subjectId)).orElse(null);
+
+			List<SubjectStudy> subjectStudyList = study.getSubjectStudyList();
+
+			for (SubjectStudy subjectStudy : subjectStudyList) {
+				if (subjectStudy.getSubject().equals(subject)) {
+					toAdd = false;
+					break;
+				} else {
+					toAdd = true;
+				}
+			}
+
+			if (toAdd) {
+				SubjectStudy ssToAdd = new SubjectStudy();
+				ssToAdd.setStudy(study);
+				ssToAdd.setSubject(subject);
+
+				subjectStudyList.add(ssToAdd);
+				study.setSubjectStudyList(subjectStudyList);
+				studyRepository.save(study);
+			}
+		}
+	}
+
+	@Override
+	public String addCenterAndCopyDatasetToStudy(String datasetIds, String id, String centerIds) {
+		String result = "";
+		Long userId = KeycloakUtil.getTokenUserId();
+		Long studyId = Long.valueOf(id);
+
+		Study study = studyService.findById(studyId);
+		StudyUser studyUser = studyUserRepository.findByUserIdAndStudy_Id(userId, studyId);
+		if (studyUser == null) {
+			LOG.error("You must be part of both studies to copy datasets.");
+			return "You must be part of both studies to copy datasets.";
+		} else {
+			List<StudyUserRight> rights = studyUser.getStudyUserRights();
+
+			if (rights.contains(StudyUserRight.CAN_ADMINISTRATE) || rights.contains(StudyUserRight.CAN_IMPORT)) {
+				addCenterToStudy(study, centerIds);
+
+				try {
+					result = copyDatasetToStudy(datasetIds, id);
+				} catch (MicroServiceCommunicationException e) {
+					throw new RuntimeException(e);
+				}
+			} else {
+				LOG.error("Missing IMPORT or ADMIN rights on destination study " + study.getName());
+				return "Missing IMPORT or ADMIN rights on destination study " + study.getName();
+			}
+			return result;
+		}
+	}
+
+	private void addCenterToStudy(Study study, String centerIds) {
+		Long studyId = study.getId();
+		for (String centerId : centerIds.split(",")) {
+			if (!centerRepository.findByStudy(studyId).contains(centerId)) {
```

Comment:
Here you are making X database calls instead of one, please change this and do a single centerRepository.findByStudy() that you'll store in a list

Reply:
Actually, study.getStudyCenterList(); get the information you need, you don't have to make this call.